### ISSUE TITLE
[Function] Storage Management (external pref / internal)

### ISSUE THREAD

**Issue by chianger** [mozilla-tw/Zerda#48]
acceptance criteria:
1. (P1) support external storage for downloads
2. (P2) support external storage for cache

testing depending on download function (#34)

**Comment by walkingice** [comment]
```diff
From 46ae8c9c7474131ff3f8fa8cee54743a4b7cf441 Mon Sep 17 00:00:00 2001
From: Julian_Chu <walkingice0204@gmail.com>
Date: Wed, 9 Aug 2017 13:33:32 +0800
Subject: [PATCH] Add a BroadcastReceiver to listen download complete event

This receiver is to copy downloaded file to removable device.

NOTICE: the implementation code is very quick and dirty implementation,
it might miss lots of error checking. As you know, file operation has
lots of exception.
---
 app/src/main/AndroidManifest.xml                   |  7 +++
 .../org/mozilla/focus/utils/FoobarReceiver.java    | 55 ++++++++++++++++++++++
 2 files changed, 62 insertions(+)
 create mode 100644 app/src/main/java/org/mozilla/focus/utils/FoobarReceiver.java

diff --git a/app/src/main/AndroidManifest.xml b/app/src/main/AndroidManifest.xml
index a6d2f78..c1f8a5b 100644
--- a/app/src/main/AndroidManifest.xml
+++ b/app/src/main/AndroidManifest.xml
@@ -83,13 +83,20 @@
         </service>
 
         <meta-data android:name="android.webkit.WebView.MetricsOptOut"
                    android:value="true" />
 
         <provider
             android:name=".provider.HistoryProvider"
             android:authorities="${applicationId}.provider.historyprovider"
             android:enabled="true"
             android:exported="false" />
+
+        <receiver android:name=".utils.FoobarReceiver">
+            <intent-filter>
+                <action android:name="android.intent.action.DOWNLOAD_COMPLETE"/>
+            </intent-filter>
+        </receiver>
+
     </application>
 
 </manifest>
\ No newline at end of file
diff --git a/app/src/main/java/org/mozilla/focus/utils/FoobarReceiver.java b/app/src/main/java/org/mozilla/focus/utils/FoobarReceiver.java
new file mode 100644
index 0000000..51ff3f2
--- /dev/null
+++ b/app/src/main/java/org/mozilla/focus/utils/FoobarReceiver.java
@@ -0,0 +1,55 @@
+/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.utils;
+
+import android.app.DownloadManager;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.database.Cursor;
+import android.net.Uri;
+import android.text.TextUtils;
+
+import org.mozilla.focus.web.Download;
+
+import java.io.File;
+
+public class FoobarReceiver extends BroadcastReceiver {
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        long downloadId = intent.getLongExtra(DownloadManager.EXTRA_DOWNLOAD_ID, -1);
+
+        DownloadManager downloadManager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
+        DownloadManager.Query query = new DownloadManager.Query();
+
+        query.setFilterById(downloadId);
+        Cursor cursor = downloadManager.query(query);
+
+        if (cursor.moveToFirst()) {
+            final int downloadStatus = cursor.getInt(cursor.getColumnIndex(DownloadManager.COLUMN_STATUS));
+            final String localUriStr = cursor.getString(cursor.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI));
+            if ((downloadStatus == DownloadManager.STATUS_SUCCESSFUL)
+                    && !TextUtils.isEmpty(localUriStr)) {
+
+                Uri localUri = Uri.parse(localUriStr);
+                File src = new File(localUri.getPath());
+                try {
+                    File outputDir = StorageUtils.getTargetDirOnRemovableStorage(context, Download.TYPE_IMAGE);
+                    if (outputDir != null) {
+                        FileUtils.ensureDir(outputDir);
+                        FileUtils.copy(src, new File(outputDir, src.getName()));
+                    }
+                } catch (NoRemovableStorageException e) {
+                    // should prompt a toast
+                    e.printStackTrace();
+                } catch (Exception e) {
+                    e.printStackTrace();
+                }
+            }
+        }
+        cursor.close();
+    }
+}
-- 
2.10.2
```

**Comment by walkingice** [comment]
I use the attached patch for test on Sony Xperia ZL.

In my case now it could copy downloaded file to 
```
shell@C6502:/storage/sdcard1/Android/media/org.mozilla.zerda.debug.jchu/downloads/pictures
```
Before integration in next sprint, I think we could close this issue since its basic function works.

**Comment by walkingice** [comment]
And I was aware of that, the **DOWNLOAD_COMPLETE** intent is not fired immediately once the download process completed. In my test (by Xperia ZL) there might be 3 to 20 seconds delay. :-/

**Comment by walkingice** [comment]
should be fixed in #443

**Comment by Conjuror** [comment]
build 972
